packages/riverpod_lint/lib/src/assists/convert_to_stateless_base_widget.dart

```diff
@@ -147,89 +150,252 @@ class ConvertToStatelessBaseWidget extends RiverpodAssist {
+        if (field != null) {
+          fieldsAssignedInConstructors.add(field);
+        }
+      }
     }
 
-    final buffer = StringBuffer();
-    final refParamStartOffset = parameters.parameters.firstOrNull?.end ??
-        parameters.leftParenthesis.offset + 1;
-
-    buffer
-      ..write(
-        source.contents.data.substring(buildMethod.offset, refParamStartOffset),
-      )
-      ..write(', WidgetRef ref')
-      ..writeln(
-        source.contents.data.substring(refParamStartOffset, buildMethod.end),
-      );
+    if (node.parent is ConstructorFieldInitializer) {
+      final element = node.staticElement;
+      if (element is FieldElement) {
+        fieldsAssignedInConstructors.add(element);
+      }
+    }
 
-    return buffer.toString();
+    if (node.inSetterContext()) {
+      final element = node.writeOrReadElement;
+      if (element is PropertyAccessorElement) {
+        final field = element.variable;
+        if (field is FieldElement) {
+          fieldsAssignedInConstructors.add(field);
+        }
+      }
+    }
   }
+}
+
+class _ReplacementEditBuilder extends RecursiveAstVisitor<void> {
+  _ReplacementEditBuilder(
+    this.widgetClassElement,
+    this.elementsToMove,
+  );
 
-  String? _buildMethodWithoutRef(MethodDeclaration buildMethod, Source source) {
-    final parameters = buildMethod.parameters;
-    if (parameters == null) return null;
+  final ClassElement widgetClassElement;
+  final Set<Element> elementsToMove;
 
-    if (parameters.parameters.length == 1) {
-      // The build method already has not a ref parameter, nothing to change
-      return '${source.contents.data.substring(buildMethod.offset, buildMethod.end)}\n';
-    }
+  List<SourceRange> deleteRanges = [];
 
-    final buffer = StringBuffer();
-    final contextEndOffset = parameters.parameters.firstOrNull?.end ??
-        parameters.leftParenthesis.offset + 1;
-    final refParamStartOffset = parameters.parameters.last.offset;
-
-    buffer
-      ..write(
-        source.contents.data.substring(buildMethod.offset, contextEndOffset),
-      )
-      ..writeln(
-        source.contents.data.substring(refParamStartOffset, buildMethod.end),
-      );
+  @override
+  void visitSimpleIdentifier(SimpleIdentifier node) {
+    if (node.inDeclarationContext()) {
+      return;
+    }
+    final element = node.staticElement;
+    if (element is ExecutableElement &&
+        element.enclosingElement == widgetClassElement &&
+        !elementsToMove.contains(element)) {
+      final parent = node.parent;
+      if (parent is PrefixedIdentifier) {
+        final grandParent = parent.parent;
+
+        if (!node.name.contains(r'$') &&
+            grandParent is InterpolationExpression &&
+            grandParent.leftBracket.type ==
+                TokenType.STRING_INTERPOLATION_EXPRESSION) {
+          final offset = grandParent.rightBracket?.offset;
+
+          if (offset != null) {
+            deleteRanges.add(SourceRange(offset, 1));
```

Comment:
```suggestion
          if (grandParent.rightBracket?.offset case final rightBracketOffset?) {
            deleteRanges.add(SourceRange(rightBracketOffset, 1));
```

Reply:
Here's a useful pattern to avoid defining local variables that are used only in a `if`.